diagnostic_rt_qpcr/operation_types/prepare_rt_qpcr_plate/protocol.rb

```diff
@@ -1,112 +1,188 @@
+
+      microtiter_plate = MicrotiterPlateFactory.build(
+        collection: output_collection,
+        group_size: group_size,
+        method: :sample_layout
+      )
+
+      composition = PCRCompositionFactory.build(
+        program_name: program_name
+      )
+      volume = { qty: composition.template.qty,
+                 units: composition.template.units }
+
+      remaining_inputs = add_control_samples(
+        operation_inputs: op.input_array(TEMPLATE),
+        microtiter_plate: microtiter_plate,
+        volume: volume
+      )
+
+      add_diagnostic_samples(
+        operation_inputs: remaining_inputs,
+        microtiter_plate: microtiter_plate,
+        volume: volume
+      )
+
+      seal_plate(output_collection)
+      show_result(collection: output_collection) if debug
     end
 
     operations.store
 
     {}
+  end
+
+  # Provides instructions and handling for addition of control
+  # samples.  Returns all inputs that are NOT control inputs
+  #
+  # @param operation_inputs [Array<item>]
+  # @param collection [Collection]
+  # @param layout_generator [LayoutGenerator]
+  # @param volume [{aty: int, unit: string}]
+  # @return operation_inputs [Array<item>]
+  def add_control_samples(operation_inputs:, microtiter_plate:,
+                          volume:)
+    %w[negative_template positive_template].each do |stub|
+      name = @operation_params["#{stub}_control".to_sym]
+      loc = @operation_params["#{stub}_location".to_sym]
+
+      next unless name.present? && loc.present?
 
+      control_inputs, operation_inputs = operation_inputs.partition do |fv|
+        fv.sample.name == name
+      end
+
+      add_samples(
+        operation_inputs: control_inputs,
+        microtiter_plate: microtiter_plate,
+        volume: volume,
+        column: loc[1]
+      )
+
+      # negative control samples need to be covered after addition
+      next unless stub.include?('negative')
+
+      seal_plate(collection, rc_list: get_rna_samples(collection))
+    end
+    operation_inputs
   end
+
+  # Provides instructions to add diagnostic samples to collection
+  #
+  # @param operation_inputs [Array<items>]
+  # @param collection [Collection]
+  # @param layout_generator [LayoutGenerator]
+  # @param volume [{aty: int, unit: string}]
+  def add_diagnostic_samples(operation_inputs:, microtiter_plate:,
+                            volume:)
+    add_samples(
+      operation_inputs: operation_inputs,
+      microtiter_plate: microtiter_plate,
+      volume: volume
+    )
+  end
+
+  # validates operations and ensures that they are formatted as expected
+  #
+  # @param operations [OperationList]
+  def validate(operations:)
+    operations.each do |op|
+      if op.input_array(TEMPLATE).length > @job_params[:max_inputs]
+        raise IncompatibleInputsError, "Too many inputs for Operation #{op.id}"
+      end
+    end
+  rescue IncompatibleInputsError => e
+    error_operations(operations: [op], err: e)
```

Comment:
This line doesn't work anymore because `op` is a local variable and would be out of scope here. The original intent was to error everything if anything was amiss. That may be too radical, but it is probably safer for now. Revert to
```
error_operations(operations: operations, err: e)
```